board/main.c

```diff
@@ -218,26 +273,58 @@ void usb_cb_ep2_out(void *usbdata, int len, bool hardwired) {
   }
 }
 
+usb_asm_buffer ep3_buffer = {.ptr = 0, .tail_size = 0, .counter = 0};
+
 // send on CAN
 void usb_cb_ep3_out(void *usbdata, int len, bool hardwired) {
   UNUSED(hardwired);
-  int dpkt = 0;
-  uint32_t *d32 = (uint32_t *)usbdata;
-  for (dpkt = 0; dpkt < (len / 4); dpkt += 4) {
-    CAN_FIFOMailBox_TypeDef to_push;
-    to_push.RDHR = d32[dpkt + 3];
-    to_push.RDLR = d32[dpkt + 2];
-    to_push.RDTR = d32[dpkt + 1];
-    to_push.RIR = d32[dpkt];
-
-    uint8_t bus_number = (to_push.RDTR >> 4) & CAN_BUS_NUM_MASK;
-    can_send(&to_push, bus_number, false);
+  uint8_t *usbdata8 = (uint8_t *)usbdata;
+  // Got first packet from a stream, resetting buffer and counter
+  if (usbdata8[0] == 0) {
+    ep3_buffer.counter = 0;
+    ep3_buffer.ptr = 0;
+    ep3_buffer.tail_size = 0;
+  }
+  // Assembling can message with data from buffer
+  if (usbdata8[0] == ep3_buffer.counter) {
+    uint8_t pos = 1;
+    ep3_buffer.counter++;
+    if (ep3_buffer.ptr != 0) {
+      if (ep3_buffer.tail_size <= 63U) {
+        CANPacket_t to_push;
+        (void)memcpy(&ep3_buffer.data[ep3_buffer.ptr], &usbdata8[pos], ep3_buffer.tail_size);
+        (void)memcpy(&to_push, ep3_buffer.data, ep3_buffer.ptr + ep3_buffer.tail_size);
+        can_send(&to_push, to_push.bus, false);
+        pos += ep3_buffer.tail_size;
+        ep3_buffer.ptr = 0;
+        ep3_buffer.tail_size = 0;
+      } else {
+        (void)memcpy(&ep3_buffer.data[ep3_buffer.ptr], &usbdata8[pos], len - pos);
+        ep3_buffer.tail_size -= 63U;
+        ep3_buffer.ptr += 63U;
+        pos += 63U;
+      }
+    }
+
+    while (pos < len) {
+      uint8_t pckt_len = CANPACKET_HEAD_SIZE + dlc_to_len[(usbdata8[pos] >> 4U)];
+      if ((pos + pckt_len) <= (uint8_t)len) {
+        CANPacket_t to_push;
+        (void)memcpy(&to_push, &usbdata8[pos], pckt_len);
+        can_send(&to_push, to_push.bus, false);
+        pos += pckt_len;
+      } else {
+        (void)memcpy(ep3_buffer.data, &usbdata8[pos], len - pos);
+        ep3_buffer.ptr = len - pos;
+        ep3_buffer.tail_size = pckt_len - ep3_buffer.ptr;
+        pos += ep3_buffer.ptr;
+      }
+    }
   }
 }
 
 void usb_cb_ep3_out_complete(void) {
-  // TODO: how does a second USB packet sneek in? (why multiply by 2)
-  if (can_tx_check_min_slots_free(MAX_CAN_MSGS_PER_BULK_TRANSFER * 2U)) {
+  if (can_tx_check_min_slots_free(MAX_CAN_MSGS_PER_BULK_TRANSFER)) {
```

Comment:
this is fixed too?

Reply:
Still testing, will ask @greg also to check, but should be fixed

Reply:
It seems like since we set the USB RX FIFO size `GRXFSIZ` to 256 bytes (64 words) it can't possibly buffer more than that many bytes.  However, `MAX_CAN_MSGS_PER_BULK_TRANSFER` currently is 64 but you can only pack 51.2 messages (min message size is 5 bytes) into 256 bytes so the current value is probably higher than necessary.